Add support for simple Web Extension tabs APIs.

xeenon opened the pull request:
#### 4f73dc6538b368a37e5b18ed205910a28658c988
<pre>
Add support for simple Web Extension tabs APIs.
<a href="https://bugs.webkit.org/show_bug.cgi?id=261412">https://bugs.webkit.org/show_bug.cgi?id=261412</a>
rdar://problem/115291352

Reviewed by Brian Weinstein.

Support for tabs.reload(), goBack(), goForward(), detectLanguage(), toggleReaderMode(),
setZoom(), getZoom(), and remove().

* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::localeStringInWebExtensionFormat): Deleted.
* Source/WebKit/Shared/Extensions/WebExtensionTab.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionTabIdentifier.h:
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.h:
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
(WebKit::toWebAPI): Added. Moved localeStringInWebExtensionFormat from CocoaHelpers.
Since this is a Web Extension specific function it makes more since here.
* Source/WebKit/Shared/Extensions/WebExtensionWindowIdentifier.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(toAPI): Use WindowVector and TabMapValueIterator from WebExtensionContext.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm:
(WebKit::WebExtensionContext::tabsReload): Added.
(WebKit::WebExtensionContext::tabsGoBack): Added.
(WebKit::WebExtensionContext::tabsGoForward): Added.
(WebKit::WebExtensionContext::tabsDetectLanguage): Added.
(WebKit::WebExtensionContext::tabsToggleReaderMode): Added.
(WebKit::WebExtensionContext::tabsGetZoom): Added.
(WebKit::WebExtensionContext::tabsSetZoom): Added.
(WebKit::WebExtensionContext::tabsRemove): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::getWindow): Use page identifier to find the tab.
(WebKit::WebExtensionContext::getTab): Added. Use page identifier to lookup active tab in window
if a tab identifier isn&apos;t specified.
(WebKit::WebExtensionContext::openWindows const): Made const and fixed a typo.
(WebKit::WebExtensionContext::openWindows): Deleted.
(WebKit::WebExtensionContext::openTabs): Deleted. Made inline in the header.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::openTabs const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPILocalizationCocoa.mm:
(WebKit::WebExtensionAPILocalization::getUILanguage):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
(WebKit::WebExtensionAPITabs::remove): Implemented. Send message to the UIProcess.
(WebKit::WebExtensionAPITabs::reload): Implemented. Ditto.
(WebKit::WebExtensionAPITabs::goBack): Implemented. Ditto.
(WebKit::WebExtensionAPITabs::goForward): Implemented. Ditto.
(WebKit::WebExtensionAPITabs::getZoom): Implemented. Ditto.
(WebKit::WebExtensionAPITabs::setZoom): Implemented. Ditto.
(WebKit::WebExtensionAPITabs::detectLanguage): Implemented. Ditto.
(WebKit::WebExtensionAPITabs::toggleReaderMode): Implemented. Ditto.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITabs.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPITabs.idl: Added NeedsPage for more functions.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITabs.mm:
(TestWebKitAPI::TEST): Added tests for all added APIs.
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.h:
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionTab initWithWindow:extensionController:]): Store the extensions controller for later.
(-[TestWebExtensionTab isShowingReaderModeForWebExtensionContext:]): Added.
(-[TestWebExtensionTab toggleReaderModeForWebExtensionContext:completionHandler:]): Added.
(-[TestWebExtensionTab detectWebpageLocaleForWebExtensionContext:completionHandler:]): Added.
(-[TestWebExtensionTab reloadForWebExtensionContext:completionHandler:]): Added.
(-[TestWebExtensionTab reloadFromOriginForWebExtensionContext:completionHandler:]): Added.
(-[TestWebExtensionTab goBackForWebExtensionContext:completionHandler:]): Added.
(-[TestWebExtensionTab goForwardForWebExtensionContext:completionHandler:]): Added.
(-[TestWebExtensionTab closeForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionWindow init]): Make _tabs mutable.
(-[TestWebExtensionWindow tabs]): Added.
(-[TestWebExtensionWindow setTabs:]): Mutable copy.
(-[TestWebExtensionWindow closeTab:]): Added.

Canonical link: <a href="https://commits.webkit.org/267880@main">https://commits.webkit.org/267880@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/174b599ae0e6aea112adfad1d1209da326ce67dc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17944 "8 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18270 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18834 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19773 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16791 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18140 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21564 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18427 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18793 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18160 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18418 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15607 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20644 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/15645 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16360 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/22894 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16661 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16529 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/20761 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17093 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/14483 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16193 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20553 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2203 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16941 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1218,6 +1230,17 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
 
+RefPtr<WebExtensionTab> WebExtensionContext::getTab(WebPageProxyIdentifier webPageProxyIdentifier, std::optional<WebExtensionTabIdentifier> identifier)
+{
+    if (!identifier) {
```
I think it might be clearer to switch these conditions, and have the first block be:

`if (identifier)`...

xeenon replied:
I agree. Will do.

xeenon committed:
Add support for simple Web Extension tabs APIs.
https://bugs.webkit.org/show_bug.cgi?id=261412
rdar://problem/115291352

Reviewed by Brian Weinstein.

Support for tabs.reload(), goBack(), goForward(), detectLanguage(), toggleReaderMode(),
setZoom(), getZoom(), and remove().

* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::localeStringInWebExtensionFormat): Deleted.
* Source/WebKit/Shared/Extensions/WebExtensionTab.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionTabIdentifier.h:
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.h:
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
(WebKit::toWebAPI): Added. Moved localeStringInWebExtensionFormat from CocoaHelpers.
Since this is a Web Extension specific function it makes more since here.
* Source/WebKit/Shared/Extensions/WebExtensionWindowIdentifier.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(toAPI): Use WindowVector and TabMapValueIterator from WebExtensionContext.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm:
(WebKit::WebExtensionContext::tabsReload): Added.
(WebKit::WebExtensionContext::tabsGoBack): Added.
(WebKit::WebExtensionContext::tabsGoForward): Added.
(WebKit::WebExtensionContext::tabsDetectLanguage): Added.
(WebKit::WebExtensionContext::tabsToggleReaderMode): Added.
(WebKit::WebExtensionContext::tabsGetZoom): Added.
(WebKit::WebExtensionContext::tabsSetZoom): Added.
(WebKit::WebExtensionContext::tabsRemove): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::getWindow): Use page identifier to find the tab.
(WebKit::WebExtensionContext::getTab): Added. Use page identifier to lookup active tab in window
if a tab identifier isn't specified.
(WebKit::WebExtensionContext::openWindows const): Made const and fixed a typo.
(WebKit::WebExtensionContext::openWindows): Deleted.
(WebKit::WebExtensionContext::openTabs): Deleted. Made inline in the header.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::openTabs const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPILocalizationCocoa.mm:
(WebKit::WebExtensionAPILocalization::getUILanguage):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
(WebKit::WebExtensionAPITabs::remove): Implemented. Send message to the UIProcess.
(WebKit::WebExtensionAPITabs::reload): Implemented. Ditto.
(WebKit::WebExtensionAPITabs::goBack): Implemented. Ditto.
(WebKit::WebExtensionAPITabs::goForward): Implemented. Ditto.
(WebKit::WebExtensionAPITabs::getZoom): Implemented. Ditto.
(WebKit::WebExtensionAPITabs::setZoom): Implemented. Ditto.
(WebKit::WebExtensionAPITabs::detectLanguage): Implemented. Ditto.
(WebKit::WebExtensionAPITabs::toggleReaderMode): Implemented. Ditto.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITabs.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPITabs.idl: Added NeedsPage for more functions.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITabs.mm:
(TestWebKitAPI::TEST): Added tests for all added APIs.
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.h:
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionTab initWithWindow:extensionController:]): Store the extensions controller for later.
(-[TestWebExtensionTab isShowingReaderModeForWebExtensionContext:]): Added.
(-[TestWebExtensionTab toggleReaderModeForWebExtensionContext:completionHandler:]): Added.
(-[TestWebExtensionTab detectWebpageLocaleForWebExtensionContext:completionHandler:]): Added.
(-[TestWebExtensionTab reloadForWebExtensionContext:completionHandler:]): Added.
(-[TestWebExtensionTab reloadFromOriginForWebExtensionContext:completionHandler:]): Added.
(-[TestWebExtensionTab goBackForWebExtensionContext:completionHandler:]): Added.
(-[TestWebExtensionTab goForwardForWebExtensionContext:completionHandler:]): Added.
(-[TestWebExtensionTab closeForWebExtensionContext:completionHandler:]):
(-[TestWebExtensionWindow init]): Make _tabs mutable.
(-[TestWebExtensionWindow tabs]): Added.
(-[TestWebExtensionWindow setTabs:]): Mutable copy.
(-[TestWebExtensionWindow closeTab:]): Added.

Canonical link: https://commits.webkit.org/267880@main
